client/lib/docker.ts

```diff
@@ -0,0 +1,201 @@
+			Env: [
+				`UDEV=0`,
+				`WORKER_TYPE=qemu`,
+			],
+			AttachStdout: true,
+			AttachStderr: true,
+			Devices:[
+				{
+					PathOnHost: "/dev/net/tun",
+					PathInContainer: "/dev/net/tun"
+				},
+				{
+					PathOnHost: "/dev/kvm",
+					PathInContainer: "/dev/kvm"
+				}
+			],
+			CapAdd: [
+				"NET_ADMIN"
+			],
+			HostConfig: {
+				Privileged: true,
+				NetworkMode: `host`,
+				Mounts: Array<any>()
+			}
+		}
+		
+		// We actually need a way to generate a unique ip address to assign the virtual interface we create, to avoid conflicts when multiple jobs run on the same jenkins worker
+		let ip = (this.containerArray.length + 1).toString();
+		let env = [
+			`UDEV=0`,
+			`WORKER_TYPE=qemu`,
+			`WORKER_PORT=${workerPort}`,
+			`CORE_PORT=${corePort}`,
+			`SCREEN_CAPTURE=true`,
+			`QEMU_BRIDGE_NAME=br${corePort}`,
+			`QEMU_BRIDGE_ADDRESS=192.168.100.${ip}`,
+		]
+
+		// volume names - we make them unique to each of the containers we spawn to avoid overlap
+		let coreStorage = `core-storage-${corePort}`;
+		let reports = `reports-${corePort}`;
+		// options to mount named volumes to the containers
+		let mounts = [
+			{
+				Target:   "/data",
+				Source:   coreStorage,
+				Type:     "volume", 
+				ReadOnly: false
+			},
+			{
+				Target:   "/reports",
+				Source:   reports,
+				Type:     "volume", 
+				ReadOnly: false
+			}
+		]
+
+		// Fill in the remaining options - we do that here as we need the port name for the env variables + volume name
+		coreOpts.Env = env;
+		coreOpts.HostConfig.Mounts = mounts;
+		workerOpts.Env = env;
+		workerOpts.HostConfig.Mounts = mounts;
+
+		await this.docker.createVolume(coreStorage);
+		await this.docker.createVolume(reports);
+		let core = await this.createContainer(`${__dirname}/../../../core`, `core`, corePort, coreOpts);
+		let worker = await this.createContainer(`${__dirname}/../../../worker`, `worker`, workerPort, workerOpts);
+
+		this.containerArray.push(
+			{
+				core: core,
+				worker: worker,
+				volumes:[
+					coreStorage,
+					reports
+				]
+			}
+		)
+	}
+
+
+	public async teardown(){
+		try{
+			for(let container of this.containerArray){
+				console.log(`Stopping ${container.core.id}...`)
+				await container.core.stop();
+				console.log(`Removing ${container.core.id}...`)
+				await container.core.remove();
+				console.log(`Stopping ${container.worker.id}...`)
+				await container.worker.stop();
+				console.log(`Removing ${container.worker.id}...`)
+				await container.worker.remove();
+	
+				// remove volumes
+				console.log(`Removing volumes`)
+				for (let volumeName of container.volumes){
+					let volume = await this.docker.getVolume(volumeName);
+					await volume.remove({force: true});
+				}
+			}
+		} catch(e){
+			console.log(e)
+		}
+	}
+
+	private async createContainer(dir: string, name: string, port: number, opts: any) {
+		console.log(`Creating ${name} container, listening on port ${port}`);
+		let imgTag = `${name}_${port}`;
```

Comment:
```suggestion
		let imgTag = `${name}:${port}`;
```
I don't love how we are using the port as a unique identifier but I haven't come up with an alternative yet. Will let you know.